crates/ruff_cli/src/diagnostics.rs

```diff
@@ -549,6 +521,102 @@ pub(crate) fn lint_stdin(
     })
 }
 
+#[derive(Debug)]
+struct Sources<'a> {
+    /// The "type" of source code, e.g. `.py`, `.pyi`, `.ipynb`, etc.
+    source_type: PySourceType,
+    /// The "kind" of source, e.g. Python file, Jupyter Notebook, etc.
+    source_kind: SourceKind,
+    /// The contents of the source code.
+    contents: Cow<'a, str>,
+}
+
+#[derive(Debug)]
+enum SourceExtractionError {
+    /// The extraction failed due to an [`io::Error`].
+    Io(io::Error),
+    /// The extraction failed, and generated [`Diagnostics`] to report.
+    Diagnostics(Box<Diagnostics>),
+}
+
+/// Extract the lint [`Sources`] from the given file path.
+fn extract_path_sources(path: &Path) -> Result<Sources, SourceExtractionError> {
+    let source_type = PySourceType::from(path);
+
+    // Read the file from disk.
+    if source_type.is_jupyter() {
+        let notebook = notebook_from_path(path).map_err(SourceExtractionError::Diagnostics)?;
+        let contents = notebook.content().to_string();
+        let source_kind = SourceKind::Jupyter(notebook);
+        Ok(Sources {
+            source_type,
+            source_kind,
+            contents: Cow::Owned(contents),
+        })
+    } else {
+        // This is tested by ruff_cli integration test `unreadable_file`
+        let contents = std::fs::read_to_string(path).map_err(SourceExtractionError::Io)?;
+        Ok(Sources {
+            source_type,
+            source_kind: SourceKind::Python,
+            contents: Cow::Owned(contents),
+        })
+    }
+}
+
+/// Extract the lint [`Sources`] from the raw string contents, optionally accompanied by a file
+/// path indicating the path to the file from which the contents were read.
+fn extract_contents_sources<'a>(
+    contents: &'a str,
+    path: Option<&Path>,
+) -> Result<Sources<'a>, SourceExtractionError> {
+    let source_type = path.map(PySourceType::from).unwrap_or_default();
+
+    if source_type.is_jupyter() {
+        // SAFETY: Jupyter isn't the default type, so we must have a path.
+        let notebook = notebook_from_contents(contents, path.unwrap())
+            .map_err(SourceExtractionError::Diagnostics)?;
+        let contents = notebook.content().to_string();
+        let source_kind = SourceKind::Jupyter(notebook);
+        Ok(Sources {
+            source_type,
+            source_kind,
+            contents: Cow::Owned(contents),
+        })
+    } else {
+        Ok(Sources {
+            source_type,
+            source_kind: SourceKind::Python,
```

Comment:
Could we instead change `SourceKind` to: 

```
enum SourceKind<'a> {
	Python(&'a str),
	Notebook(Notebook)
}

impl SourceKind<'_> {
	fn source_code(&self) -> &str {
		match self {
			Self::Python(source) => source,
			Self::Notebook(source) => source.as_str()
		}
}
```

Reply:
I'm using `source_code`here (or `source`) because I find `contents` a too generic term (I can't even tell from its name if it is a string)

Reply:
This was my instinct too. I tried it but it didn’t work, because the Diagnostics struct includes the SourceKind so that it can reconstruct the cell ranges when it reports diagnostics at the end. So we’d have to add lifetimes to Diagnostics and then keep all the source code in memory for the life of the program.

Reply:
Hmm I see. I have to take a closer look at this tomorrow. I wonder if `SourceKind` (or something similar) should be a trait and / or if we can structure these types differently to also avoid the overlap between `SourceKind` and `SourceType`

Reply:
Diagnostics only needs access to the `index` though, not the file contents, which may help.

Reply:
Yeah, it's another source mapping (the reverse after applying fixes). It would be nice to have a more formal concept for that rather than special casing jupyter notebooks in the `Emitter`. Because we'll have the same problem when linting markdown files: We need to map back the relative code block indices to the absolute file indices, potentially customizing the message.